Plugin: Correctly classify functionality in the `lib` folder

gziolo opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Part of #39889 that lists the changes to backport for the WordPress 6.0 release.

Related Site Editor core merge commit https://github.com/WordPress/wordpress-develop/commit/f034bc832eb091079e4130f6a206b4f496766796.
Related Trac ticket for the navigation editor https://core.trac.wordpress.org/ticket/50544.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

We need to correctly classify added code changes so we know exactly whether they are still experimental, they were shipped in WordPress 5.9 or they should be backported to WordPress core as part of 6.0 major release.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

I moved files from the `lib` folder to subfolders after checking whether they are still experimental, were backported to WordPress 5.9 or I believe they should be backported to WordPress 6.0.

### Experimental

- [editor-settings.php](https://github.com/WordPress/gutenberg/tree/update/lib-triage/lib/experimental/editor-settings.php)
- [experiments-page.php](https://github.com/WordPress/gutenberg/tree/update/lib-triage/lib/experimental/experiments-page.php)
- [navigation-page.php](https://github.com/WordPress/gutenberg/tree/update/lib-triage/lib/experimental/navigation-page.php)
- [navigation-theme-opt-in.php](https://github.com/WordPress/gutenberg/tree/update/lib-triage/lib/experimental/navigation-theme-opt-in.php)
- [rest-api.php](https://github.com/WordPress/gutenberg/tree/update/lib-triage/lib/experimental/rest-api.php)
- [style-engine](https://github.com/WordPress/gutenberg/tree/update/lib-triage/lib/experimental/style-engine) folder

### WordPress 5.9

- [block-gallery.php](https://github.com/WordPress/gutenberg/tree/update/lib-triage/lib/compat/wordpress-5.9/block-gallery.php)
- [edit-site-page.php](https://github.com/WordPress/gutenberg/tree/update/lib-triage/lib/compat/wordpress-5.9/edit-site-page.php)
- [kses.php](https://github.com/WordPress/gutenberg/tree/update/lib-triage/lib/compat/wordpress-5.9/kses.php)
- [navigation.php](https://github.com/WordPress/gutenberg/tree/update/lib-triage/lib/compat/wordpress-5.9/navigation.php)
- [rest-api.php](https://github.com/WordPress/gutenberg/tree/update/lib-triage/lib/compat/wordpress-5.9/rest-api.php)

### WordPress 6.0

- [client-assets.php](https://github.com/WordPress/gutenberg/tree/update/lib-triage/lib/compat/wordpress-6.0/client-assets.php)
- [functions.php](https://github.com/WordPress/gutenberg/tree/update/lib-triage/lib/compat/wordpress-6.0/functions.php)

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

No changes at all. The plugin should work as before.

gziolo committed:
Plugin: Correctly classify functionality in the `lib` folder

gziolo committed:
Move code that should be backported to WordPress core as part of 6.0

gziolo committed:
Update CODEOWNERS file

adamziel commented on lib/compat/wordpress-5.9/rest-api.php:
```diff
@@ -0,0 +1,73 @@
+ * @param WP_REST_Response $response Response data served by the WordPress REST index endpoint.
+ * @return WP_REST_Response
+ */
+function gutenberg_register_site_logo_to_rest_index( $response ) {
```
good call adding the gutenberg_prefix

adamziel reviewed (approved):
LGTM, I restarted the failing E2E test that, again, seems to be just flaky.

gziolo commented on .github/CODEOWNERS:
```diff
@@ -132,12 +132,12 @@
 /lib/compat/wordpress-5.9/theme-i18n.json                            @timothybjacobs @spacedmonkey @oandregal
 /lib/compat/wordpress-5.9/class-wp-theme-json-gutenberg.php          @timothybjacobs @spacedmonkey @oandregal
 /lib/compat/wordpress-5.9/class-wp-theme-json-resolver-gutenberg.php @timothybjacobs @spacedmonkey @oandregal
+/lib/compat/wordpress-6.0/global-styles.php                          @timothybjacobs @spacedmonkey @oandregal
```
@oandregal, can you confirm that all changes in https://github.com/WordPress/gutenberg/blob/1db15f0072f02b6a327991c7789839d607c568f5/lib/compat/wordpress-6.0/global-styles.php should be backported to WordPress 6.0?

gziolo replied:
Got this link https://github.com/WordPress/wordpress-develop/commit/872b81879415748b3177f73db5280c6348514c21 from WP 5.9. I will update the patch.

oandregal replied:
Yeah, the remaning changes in `lib/global-styles.php` were all backported for 5.9. Given the remaining code in the file only deals with kses it can also renamed to `kses.php` for clarity.

gziolo replied:
Done in https://github.com/WordPress/gutenberg/pull/39972/commits/0a8e9c9a33577224a934fa9ddc6b064412744a86 👍🏻

gziolo committed:
Move global styles from 6.0 to 5.9 as kses.php

gziolo commented on .github/CODEOWNERS:
```diff
@@ -132,12 +132,7 @@
 
 # PHP
 /lib                                                                 @timothybjacobs @spacedmonkey
-/lib/global-styles.php                                               @timothybjacobs @spacedmonkey @oandregal
```
I don't think we need to list files that were backported to WordPress 5.9.

oandregal replied:
In this case, it has been particularly useful to have these automatic notifications to share that we shouldn't touch those files with those unaware of the process. I'd like to keep them if that's possible.

oandregal replied:
(I'm talking about the `theme.json` stuff: classes, json files)

gziolo replied:
Sure,it is. Reverting :)

gziolo commented:
@noisysocks, @talldan, and @anton-vlasenko, can you confirm whether all changes applied to [navigation.php](https://github.com/WordPress/gutenberg/blob/563e51a08ad5c7a35efa3d9920b1bf7b2f338736/lib/compat/wordpress-5.9/navigation.php) were fully backported for WordPress 5.9 release? I definitely can find some changes in this commit https://github.com/WordPress/wordpress-develop/commit/f034bc832eb091079e4130f6a206b4f496766796. The question is whether we still need to add some changes as part of WordPress 6.0.

I'd like also to confirm that all code in [navigation-theme-opt-in.php](https://github.com/WordPress/gutenberg/blob/563e51a08ad5c7a35efa3d9920b1bf7b2f338736/lib/experimental/navigation-theme-opt-in.php) is still experimental. Functions reference still open Trac ticker: https://core.trac.wordpress.org/ticket/50544, so I assumed it's still not part of WordPress core. The same question applies to [navigation-page.php](https://github.com/WordPress/gutenberg/blob/563e51a08ad5c7a35efa3d9920b1bf7b2f338736/lib/experimental/navigation-page.php).

gziolo committed:
Move more files to the compat layer

gziolo commented:
I updated more code in https://github.com/WordPress/gutenberg/pull/39972/commits/9bc946f33766487cd6d0215ac4029350c2040aa3:

- `lib/full-site-editing/edit-site-page.php` moved to `lib/compat/wordpress-5.9/edit-site-page.php` after confirming with @Mamaduka in https://github.com/WordPress/gutenberg/issues/39639#issuecomment-1086800050.
- `lib/compat.php` moved to `lib/compat/wordpress-5.9/block-gallery.php` after confirming with @glendaviesnz. I moved `gutenberg_safe_style_attrs` to `lib/compat/wordpress-5.9/blocks.php` which has an explicit note about the minimal required version but isn't targeting the Gallery block only.

gziolo committed:
Make the style engine class experimental

gziolo committed:
Move all the remaining code in lib

gziolo commented:
I'm pretty happy about the current folder structure:

<img width="208" alt="Screenshot 2022-04-04 at 10 18 43" src="https://user-images.githubusercontent.com/699132/161503187-b1fa4cda-54bd-4313-9acf-4c14cd5b743e.png">

I tried removing `pwa.php` and `service-worker.js` in #39930 but there were some surprising e2e test errors so I reverted it. I will try again next week.

With [bfa2120](https://github.com/WordPress/gutenberg/pull/39972/commits/bfa2120cea7e287c9366551877c7d2061c7f6ca0) I moved `lib/style-engine` to `lib/experimental/style-engine`.

With [7d631be](https://github.com/WordPress/gutenberg/pull/39972/commits/7d631be72e3b0bc208503cc064f527b2cac8264d) I moved `lib/rest-api.php` to `lib/experimental/rest-api.php` and extracted and moved some code to `lib/wordpress-5.9/rest-api.php`.

I also renamed `lib/wordpress-6.0/utils.php` to `lib/wordpress-6.0/functions.php` to align with the final location in WordPress core.

youknowriad commented on lib/client-assets.php:
```diff
@@ -182,36 +182,6 @@ function gutenberg_override_style( $styles, $handle, $src, $deps = array(), $ver
- *
- * @param WP_Scripts $scripts WP_Scripts instance.
- */
-function gutenberg_register_vendor_scripts( $scripts ) {
```
With React 18, we might need to move that to another folder, but it's a problem for later :)

gziolo replied:
I took that into account. It's probably fine to move it between WP versions. When `SCRIPT_DEBUG` is enabled I inject the dev scripts for React Fast Refresh so I will think how to add it to WP core so we don't have to include that every time we update the React version.

youknowriad reviewed (commented):
Really cool PR :) Thanks for doing that.

For the `edit-site-page.php` I'm not sure whether we can just move it as is to 5.9 lib. It seems there's some changes in the "preloading" in 6.0. So I guess either we move the file to 6.0 or find a way to override the preloading stuff in 6.0 folder.

gziolo commented:
> For the `edit-site-page.php` I'm not sure whether we can just move it as is to 5.9 lib. It seems there's some changes in the "preloading" in 6.0. So I guess either we move the file to 6.0 or find a way to override the preloading stuff in 6.0 folder.

There is a hook that targets only the preloading functionality. I will land this PR to help coordinate backports to WordPress core and see how we can attack the preloading in a more granular way 💯

gziolo commented:
> > For the `edit-site-page.php` I'm not sure whether we can just move it as is to 5.9 lib. It seems there's some changes in the "preloading" in 6.0. So I guess either we move the file to 6.0 or find a way to override the preloading stuff in 6.0 folder.
> 
> There is a hook that targets only the preloading functionality. I will land this PR to help coordinate backports to WordPress core and see how we can attack the preloading in a more granular way 💯

It looks like @jsnajdr has covered everything for WordPress 5.9 with this hook:

https://github.com/WordPress/gutenberg/blob/4bd51e16e00cf5bbc8ae81ed8712c6bde8a83a66/lib/compat/wordpress-6.0/edit-form-blocks.php#L14-L51

jasmussen commented:
I think this PR may have accidentally broken some of the Site Logo functionality. Before this PR, the W would be replaced by any site logo set, like so:

<img width="1392" alt="Screenshot 2022-04-04 at 17 29 06" src="https://user-images.githubusercontent.com/1204802/161585349-558cf835-2ecd-42e7-a025-3ea617b48aeb.png">

After this PR, that's broken:

<img width="1392" alt="Screenshot 2022-04-04 at 17 29 47" src="https://user-images.githubusercontent.com/1204802/161585377-3c6f2e4a-dde5-4b5d-b21f-524fd2c3979b.png">

Steps to reproduce:

* Insert the site logo block
* Add a site logo
* Save and publish
* Reload the editor

The W should be replaced by a round-rectangle site logo instead.

gziolo commented on lib/init.php:
```diff
@@ -101,50 +101,3 @@ function gutenberg_rest_nonce() {
- * @param WP_REST_Response $response Response data served by the WordPress REST index endpoint.
- * @return WP_REST_Response
- */
-function register_site_icon_url( $response ) {
-	$data                  = $response->data;
-	$data['site_icon_url'] = get_site_icon_url();
-	$response->set_data( $data );
-	return $response;
-}
-
-add_filter( 'rest_index', 'register_site_icon_url' );
```
@jasmussen, great catch. I accidentally removed this filter related to the Site Logo block 😱 

I will open PR as the first thing in the morning.

gziolo replied:
Ready in #40041.
